### ISSUE TITLE
Fail with error "trying to activate test environment of an unnamed project"

### ISSUE THREAD

**Issue by jlapeyre** [JuliaTesting/ReTestItems.jl#34]
This happens both when running `test` in `Pkg` mode at the CLI and when `using ReTestItems.jl` at the Julia CLI and then `runtests`.

For example
```julia
julia> runtests("test/dummy_tests.jl")
ERROR: trying to activate test environment of an unnamed project
```

Or when running `Pkg.test`:
```julia
ERROR: LoadError: trying to activate test environment of an unnamed project
in expression starting at /home/lapeyre/myrepos/quantum_repos/QuantumDAGS/test/runtests.jl:4
```
It appears to be when `TestEnv` calls `Context().env` and gets `nothing` instead of package info

I tried using the main branch of `ReTestItems`, and a cloned one as well, but then `test` fails with "cannot merge projects"

I am using a `test/Project.toml`.

`ReTestItems` v1.1.1
Julia 1.8.5 and 1.9.0-rc1

**Comment by nickrobinson251** [comment]
> I am using a test/Project.toml.

Unfortunately that's not supported right now.

**Comment by nickrobinson251** [comment]
put up a fix, if you'd like to try it out https://github.com/JuliaTesting/ReTestItems.jl/pull/36

**Comment by jlapeyre** [comment]
I tried #36 like this:
```julia
(test) pkg> add  ReTestItems#npr-support-test-project-files
```
But this gives me "ERROR: can not merge projects".

I've read https://github.com/JuliaLang/Pkg.jl/issues/1585 and tried some things, but I'm not sure the situation there applies in my case. IIUC they are talking about ./Manifest.toml and ./test/Manifest.toml having different paths to ReTestItems.jl. But  ./Manifest.toml does not have an entry at all for ReTestItems.jl. 

Some people there mentioned removing the package being tested from `./test/Project.toml`. But it is not there to be removed in my case.

**Comment by nickrobinson251** [comment]
is the codebase that you're trying to test open-source? i wonder if i could try it out myself to help narrow this down

I also wonder if it's worth deleting the various Manfiest.toml's are re-creating them (if that's an option for you)

**Comment by jlapeyre** [comment]
I tried deleting Manifest.toml s, adding and removing packages. There is no stack trace just an error message. Maybe Pkg.test is eating the stack trace. So this makes is hard to track down.

The package that I'm working on now is not yet opensourced, although it will be if all goes as planned. But I also tried in on an OS repo, same problem. In fact, let me try on my simplest OS package and if it fails, I'll point you there. (I'm hoping this allows us to run unit tests, aqua, and jet in parallel.)

**Comment by jlapeyre** [comment]
Here are two branches using your bugfix branch that both give identical test failure mode. The things I tried are explained in the commit messages.
The bottom line is that the error messge is "ERROR: trying to activate ..."

https://github.com/jlapeyre/ControlFlow.jl/tree/use-retest-items

https://github.com/jlapeyre/ControlFlow.jl/tree/use-retest-items-and-edit-main-projecttoml

**Comment by nickrobinson251** [comment]
Ah, yeah, i see a few issues here:
- The main issue, on both branches, is that `test/runtests.jl` is calling `runtests()` when it should call `runtests(ControlFlow)`
  - `runtests()` tries to run tests _in the current env_, but when you run `Pkg.test()` (or `] test`) _that always creates a new env_ ...and trying to run tests for the new test env leads to `trying to activate test environment of an unnamed project`
- on [use-retest-items-and-edit-main-projecttoml](https://github.com/jlapeyre/ControlFlow.jl/tree/use-retest-items-and-edit-main-projecttoml), you need to delete the test/Project.toml and test/Manifest.toml when using a `./Project.toml` with test-deps declared in `[extras]`
- on [use-retest-items](https://github.com/jlapeyre/ControlFlow.jl/tree/use-retest-items) you need to add `Test` to the `test/Project.toml`

Tests run on both branches for me with these changes

**Comment by jlapeyre** [comment]
I verify that your edits on [use-retest-items-and-edit-main-projecttoml](https://github.com/jlapeyre/ControlFlow.jl/tree/use-retest-items-and-edit-main-projecttoml) also work for me.

My best recollection is that Aqua complains that it wants test-deps in `[extras]` even with a `test/Project` for compatibility.

So this seems to solve the problem for `ControlFlow`. But it never showed me the "cannot merge" error.

Here is a just slightly more complicated package that gave me the "cannot merge project" error. There are several commits with messages about what I changed and what the  result was.

https://github.com/jlapeyre/DictTools.jl/compare/main...use-retest-items

**Comment by jlapeyre** [comment]
I  added more commits to the branch of `DictTools` above. The last commit https://github.com/jlapeyre/DictTools.jl/commit/d1e9f09ebcf891b7c6454394e4956c6ba46d76ef works to a degree. RTI finds and executes `@testitem`s in  `./src`. But not in `./test`. This is with the released branch of RTI. I am unable to get past `cannot merge` when using the dev branch.

**Comment by nickrobinson251** [comment]
i won't have time to look at this again til next week

Overall i'm inclined to say i've never seen a `test/Project.toml` work well
(and others agree it's not recommended e.g. https://github.com/JuliaLang/Pkg.jl/issues/1585#issuecomment-578187367)

**Comment by jlapeyre** [comment]
Thanks for all the effort. I haven't had too much trouble with them myself, till now. But my experience and issues online suggest it's problematic. Now that I know (I hope) it's `test/Project.toml` causing the trouble, I'll try switching to the old method.

**Comment by jlapeyre** [comment]
FWIW. If I use `[extras]` and `[targets]` instead of `test/Project.toml` then `runtests` seems to work as advertised. So whatever caused the "cannot merge" does not seem to affect the older way to run tests for me. In any case it's starting to look like the focus of `ReTestItems` is to use it outside of the standard test framework.

I'd be happy if the developers of `ReTestItems` wanted to work on other things until/if using `test/Project.toml`  is less buggy.

**Comment by nickrobinson251** [comment]
closing as won't fix
